### PR TITLE
implement SSE support

### PR DESCRIPTION
**Pull request by xxchan**
part of https://github.com/aos-dev/go-storage/issues/523

ref: https://docs.qingcloud.com/qingstor/api/common/encryption.html

### CODE REVIEW

File: service.toml [hunk 1]
```diff
@@ -36,10 +36,32 @@ optional = ["list_mode"]
...
 
 [namespace.storage.op.write]
-optional = ["content_md5", "content_type", "io_callback", "storage_class"]
+optional = ["content_md5", "content_type", "io_callback", "storage_class", "sse_customer_algorithm", "sse_customer_key", "sse_customer_key_md5"]
+
+[namespace.storage.op.copy]
+optional = ["sse_customer_algorithm", "sse_customer_key", "sse_customer_key_md5", "sse_copy_source_customer_algorithm", "sse_copy_source_customer_key", "sse_copy_source_customer_key_md5"]
+
+[pairs.sse_customer_algorithm]
+type = "string"
```
Comment (Xuanwo) [inline service.toml:48]: Maybe we can provide `sse_customer_algorithm` in `enum`?

File: service.toml [hunk 2]
```diff
@@ -36,10 +36,32 @@ optional = ["list_mode"]
...
+
+[pairs.sse_customer_key]
```
Comment (Xuanwo) [inline service.toml:50]: Let's add some `description` here, how to construct a valid key? What value we need here?

File: service.toml [hunk 3]
```diff
@@ -36,10 +36,32 @@ optional = ["list_mode"]
...
+type = "string"
+
+[pairs.sse_customer_key_md5]
```
Comment (Xuanwo) [inline service.toml:53]: ditto

File: storage.go [hunk 4]
```diff
@@ -349,7 +355,11 @@ func (s *Storage) reach(ctx context.Context, path string, opt pairStorageReach)
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	input := &service.GetObjectInput{}
+	input := &service.GetObjectInput{
+		XQSEncryptionCustomerAlgorithm: &opt.SseCustomerAlgorithm,
```
Comment (Xuanwo) [inline storage.go:359]: We need to check value before use them, please check `HasSseCustomerAlgorithm` first.

File: service.toml [hunk 5]
```diff
@@ -36,10 +36,32 @@ optional = ["list_mode"]
...
+
+[pairs.sse_copy_source_customer_algorithm]
+type = "string"
+
+[pairs.sse_copy_source_customer_key]
+type = "string"
+
+[pairs.sse_copy_source_customer_key_md5]
+type = "string"
+
```
Comment (Xuanwo) [inline service.toml:64]: We also need to add an obejct info meta to carry `x-qs-encryption-customer-algorithm`'s value in object.

**Comment by xxchan** [comment]
PTAL @Xuanwo

File: storage.go [hunk 6]
```diff
@@ -421,6 +441,11 @@ func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int6
 	if opt.HasStorageClass {
 		input.XQSStorageClass = service.String(opt.StorageClass)
 	}
+	if opt.HasSseCustomerAlgorithm {
```
Comment (Xuanwo) [inline storage.go:444]: `Multiparter` related operations should also add SSE-related pairs?

**Comment by Xuanwo** [comment]
Let's add a pairs section in https://aos.dev/docs/go-storage/services/qingstor

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**